drivers/interrupt_controller/intc_plic.c

```diff
@@ -208,43 +298,75 @@ static void plic_irq_handler(const void *arg)
 /**
  * @brief Initialize the Platform Level Interrupt Controller
  *
+ * @param dev PLIC device struct
+ *
  * @retval 0 on success.
  */
-static int plic_init(void)
+static int plic_init(PLIC_INIT_FN_ARG(dev))
 {
-
-	volatile uint32_t *en = (volatile uint32_t *)PLIC_IRQ_EN;
-	volatile uint32_t *prio = (volatile uint32_t *)PLIC_PRIO;
-	volatile struct plic_regs_t *regs =
-	    (volatile struct plic_regs_t *)PLIC_REG;
-	int i;
+#ifdef CONFIG_XIP
+	const struct device *dev = NULL;
+#endif
+	const struct plic_config *config = get_plic_config(dev);
+	volatile uint32_t *en = (volatile uint32_t *) config->irq_en;
+	volatile uint32_t *prio = (volatile uint32_t *) config->prio;
+	volatile struct plic_regs_t *regs = (volatile struct plic_regs_t *) config->reg;
 
 	/* Ensure that all interrupts are disabled initially */
-	for (i = 0; i < PLIC_EN_SIZE; i++) {
+	for (uint32_t i = 0; i < get_plic_enabled_size(dev); i++) {
 		*en = 0U;
 		en++;
 	}
 
 	/* Set priority of each interrupt line to 0 initially */
-	for (i = 0; i < PLIC_IRQS; i++) {
+	for (uint32_t i = 0; i < config->num_irqs; i++) {
 		*prio = 0U;
 		prio++;
 	}
 
 	/* Set threshold priority to 0 */
 	regs->threshold_prio = 0U;
 
-	/* Setup IRQ handler for PLIC driver */
-	IRQ_CONNECT(RISCV_MACHINE_EXT_IRQ,
-		    0,
-		    plic_irq_handler,
-		    NULL,
-		    0);
-
-	/* Enable IRQ for PLIC driver */
-	irq_enable(RISCV_MACHINE_EXT_IRQ);
+	/* Configure IRQ for PLIC driver */
+	config->irq_config_func();
 
 	return 0;
 }
 
+#define PLIC_INTC_IRQ_FUNC_DECLARE(n) static void plic_irq_config_func_##n(void)
+
+#define PLIC_INTC_IRQ_FUNC_DEFINE(n)                                                               \
+	static void plic_irq_config_func_##n(void)                                                 \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(n), 0, plic_irq_handler,                                  \
+			    COND_CODE_1(CONFIG_XIP, (NULL), (DEVICE_DT_INST_GET(n))), 0);          \
```

Comment:
Not 100% sure if we need the cond_code_1 and config_xip conditionals here.. but otherwise, the looks fine

Reply:
this was required to workaround the XIP issue, the XIP workaround commit has been removed for now